x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/InferenceMetadataFieldMapper.java

```diff
@@ -0,0 +1,427 @@
+                    SemanticTextModelSettings.INFERENCE_ID_FIELD.getPreferredName(),
+                    modelSettings.inferenceId(),
+                    original.name(),
+                    SemanticTextModelSettings.INFERENCE_ID_FIELD.getPreferredName(),
+                    modelSettings.inferenceId()
+                )
+            );
+        }
+        if (modelSettings.taskType() == TaskType.TEXT_EMBEDDING && modelSettings.dimensions() == null) {
+            throw new DocumentParsingException(
+                xContentLocation,
+                "Model settings for field [" + original.name() + "] must contain dimensions"
+            );
+        }
+        if (original.getModelSettings() == null) {
+            if (parent != docContext.root()) {
+                mapperBuilderContext = mapperBuilderContext.createChildContext(parent.name(), ObjectMapper.Dynamic.FALSE);
+            }
+            SemanticTextFieldMapper newMapper = new SemanticTextFieldMapper.Builder(
+                original.simpleName(),
+                docContext.indexSettings().getIndexVersionCreated(),
+                docContext.indexAnalyzers()
+            ).setInferenceId(modelSettings.inferenceId()).setModelSettings(modelSettings).build(mapperBuilderContext);
+            docContext.addDynamicMapper(newMapper);
+            return newMapper.getSubMappers();
+        } else {
+            SemanticTextFieldMapper.Conflicts conflicts = new Conflicts(original.name());
+            SemanticTextFieldMapper.canMergeModelSettings(original.getModelSettings(), modelSettings, conflicts);
+            try {
+                conflicts.check();
+            } catch (Exception exc) {
+                throw new DocumentParsingException(xContentLocation, "Incompatible model_settings", exc);
+            }
+        }
+        return original.getSubMappers();
+    }
+
+    private void parseSingleField(DocumentParserContext context, MapperBuilderContext mapperBuilderContext) throws IOException {
+        XContentParser parser = context.parser();
+        String fieldName = parser.currentName();
+        var res = findMapper(context.mappingLookup().getMapping().getRoot(), fieldName);
+        if (res == null || res.mapper == null || res.mapper instanceof SemanticTextFieldMapper == false) {
+            throw new DocumentParsingException(
+                parser.getTokenLocation(),
+                Strings.format("Field [%s] is not registered as a [%s] field type", fieldName, SemanticTextFieldMapper.CONTENT_TYPE)
+            );
+        }
+        parser.nextToken();
+        failIfTokenIsNot(parser.getTokenLocation(), parser, XContentParser.Token.START_OBJECT);
+
+        // record the location of the inference field in the original source
+        XContentLocation xContentLocation = parser.getTokenLocation();
+        // parse eagerly to extract the model settings first
+        Map<String, Object> map = parser.mapOrdered();
```

Comment:
I don't think it's an issue, we do parse the results twice but we do it per field so it won't take a lot of memory.
So unless we change the structure of the object we'll need to do this double parsing since we cannot rely on the order of the field in a json map.